### ISSUE TITLE
Add `@upcast` and `@downcast`

### ISSUE THREAD

**Issue by AndrolGenhald** [vimeo/psalm#7675]
`@var` is dangerous because it overrides the type and will give no warnings if the inferred type changes to be incompatible.

I see two use-cases for `@var` that could use additional checks: upcasting and downcasting.

Downcasting is often used when a return type is known to satisfy some Psalm specific type but it's not stubbed. If we add `@downcast` as an alternative to `@var` we can add a check to ensure that the type being downcast from is actually a supertype of the new type: https://psalm.dev/r/6a51c65115 https://psalm.dev/r/282f335678 https://psalm.dev/r/2954ff83d0

Upcasting is sometimes needed for templates since the inferred type may be narrower than the desired type: https://psalm.dev/r/fa6808b91e https://psalm.dev/r/e2ce2be20f https://psalm.dev/r/4d023f6541 https://psalm.dev/r/eec78b915f
The upcast check is slightly more complicated because it needs to allow upcasting invariant templates, even though the type being cast from is not a subtype of the type being cast to: https://psalm.dev/r/70792294b9

**Comment by ondrejmirtes** [comment]
My personal opinion is that there's no need for these and that any new PHPDoc tag will have a hard task of finding any adoption, making it less interesting for me to be implemented in PHPStan.

There's no need for an inline `@var` nor these new tags. All of these problems can be solved with:

* Stub files - for overriding wrong 3rd party PHPDocs
* Generics, conditional types, and similar facilities - for specifying what the return type really is

What remains is telling the IDE what to autocomplete, which should be solved by opening bugreports and demanding feature parity with SA tools.

What I recommend SA users is an interesting exercise - try to remove all inline `@var` from your codebase and try to live without them.

**Comment by orklah** [comment]
Yeah, but if we're honest for a minute, generating stubs is a tool for power users... I maintain Psalm and I don't even bother doing this for my work codebase.

Sure, it's way cleaner, but most people use SA tools to have less bugs, not because they obsess with cleanliness.

@var is clearly a big issue, mainly because it's ambiguous. It serves multiples purposes for multiple tools and for SA, it's quite a powerful tool with no safety so we need to phase it out. But I don't see myself removing it without offering something else in return...

**Comment by AndrolGenhald** [comment]
@ondrejmirtes agreed for the most part, I could maybe do without `@downcast`, but for now [this](https://phpstan.org/r/1dec50f1-4077-467f-9c74-829f6d74d063) is broken in both Psalm and PHPStan and it seems like the only way to fix it currently is to use a `@var` tag (you could change the code too of course, but that could be complicated).

I might try to get this sort of upcasting to happen automatically in Psalm, but until/unless that works I think an `@upcast` tag is useful.

**Comment by janopae** [comment]
> agreed for the most part, I could maybe do without @downcast, but for now [this](https://phpstan.org/r/1dec50f1-4077-467f-9c74-829f6d74d063) is broken in both Psalm and PHPStan and it seems like the only way to fix it currently is to use a @var tag

This is true, I just created an issue for this: https://github.com/vimeo/psalm/issues/9838

But the scope of this is way smaller than introducing an `@upcast` annotation. It's just that automatic type inference doesn't work in these cases. I think that's why automatic type inference is optional in languages like Java, and you can always provide the types manually when creating an object.

The `@upcast` annotation has the problem that you could use it to change the type lateron. This could lead to [Dog collections filled with Cats](https://psalm.dev/docs/annotating_code/templated_annotations/#template-covariance). We need an annotation that does nothing more than letting you specify the template types on the creation of an object (like the `@template-params` annotation I proposed in my issue).

This would solve the fact that there is no way to properly type these cases without using a `@var` annotation in both PHPStan and Psalm, and it wouldn't lead to undesired consequences.

@ondrejmirtes would you be okay with such an annotation? This might lead to people actually being able to remove all inline `@var` from their codebases and live without them.
